### ISSUE TITLE
Explore alternative API for websocket server programming

### ISSUE THREAD

**Issue by jedesah** [http4s/http4s#316]
As talked about in #311, it is unfortunate that an echo server requires a `Queue` and I have spent a little time to explore whether a better API might be possible.

As suggested by @rossabaker, I tried modifying the [chat server in the demo application](https://github.com/http4s/http4s_demo/blob/b8b4781aa904ec65a37f029cfb6e1adc7118837a/src/main/scala/org/http4s/example/Routes.scala#L33-L47) to fit this new model and see what it looks like. This is what I came up with:

    case r @ GET -> Root / "wschat" / name =>

      def frameToMsg(f: WebSocketFrame) = f match {
        case Text(msg, _) => s"$name says: $msg"
        case _            => s"$name sent bad message! Booo..."
      }

      chatTopic.publishOne(s"New user '$name' joined chat")
      .flatMap {_ => WS { in =>

        val receiveMsgs = in.map(frameToMsg).onComplete(Process.eval_(chatTopic.publishOne(s"$name left the chat"))).to(chatTopic.publish)
        receiveMsgs.run.runAsync(_ => ())
        Process.emit(Text(s"Welcome to the chat, $name!")) ++ chatTopic.subscribe.map(Text(_))
      }

The rest is unchanged.

Unfortunately, this requires calling `runAsync` which is not ideal, but I don't think there is a better way. The problem stems from the fact that the `in` and `out` streams are not really connected. In the case of an echo server, this model shines because the framework will run the out stream in order to send the messages to the client and that will consume the messages from the client. But in our case, without running the in process in order to pipe it's results out to `chatTopic.publish`, the out stream will simply indefinitely wait on someone to publish, but no one will be running in order to do so.

I tend to still prefer this solution because it is simple in the straightforward case where input is being mapped in some way to some output. Even though `topic` is a really cool construct, it is in the `async` package for a reason. Like `Queue` it is using mutability to accomplish things and as such I would expect a user of a `Queue` or a `Topic` to be aware that the `input` stream needs to be run somehow in oder to actually be mutating either the `Queue` or the `Topic`.

Thoughts?

**Comment by bryce-anderson** [comment]
I think we should pull a streams expert into the discussion. @djspiewak comes to mind as a http4s user and stream contributor, I hope he doesn't mind being called out :smile:. 

I tend to agree that the ```Process => Process``` signature would be ideal for situations where the server just receives a series of messages for which each gets a response. However needing to do an ```awaitAsync``` is a red flag in my mind: we really want the backend to be in charge of driving the execution. What we could do to avoid this is use a ```wye``` to merge the read process into the write process and cause the backend to drive both the send and receive but that is a very large step back from the `Exchange` interface. This is just speculation: I haven't tried it, maybe it doesn't actually work, maybe its awesome.

Another subjective question is what solution is most _idiomatic_, and I'm going to guess `Exchange`. The [nio](https://github.com/scalaz/scalaz-stream/blob/master/src/main/scala/scalaz/stream/nio/package.scala) package used an `Exchange` as the interface for a socket which is really what we are modeling with a websocket.

Hopefully we can get some more expert opinions on the matter.

**Comment by jedesah** [comment]
> I tend to agree that the Process => Process signature would be ideal for situations where the server just receives a series of messages for which each gets a response.

It doesn't need to be that each and every message that comes in needs to have an equivalent response going out. It just needs to be that the output stream is pulling from the input stream somehow.

New idea just in, maybe the ideal solution is to have a constructor for `Exchange` that takes a `Process => Process`, let me have a look at that.

What I like about `Exchange` is that you can't really go wrong, it basically says give me a stream to run that will produce what I need to send to the client and give me a sink I can shove the client messages down. What I don't like is that I believe it will lead people to write these huge sinks that are essentially a big effectful function. On top of not being very composable, it will lead to bizarre patterns in order to hook up the input to the output. As demonstrated with the echo server, I don't think it is really possible to accomplish that without using a mutable Queue.

I look forward to hearing what Daniel thinks about this all of this.

**Comment by jedesah** [comment]
This confuses me: 
```
    * @param read Process reading values from remote system
    * @param write Process writing values to remote system
    *
    * @tparam I  values read from remote system
    * @tparam W  values written to remote system
    */
    final case class Exchange[I, W](read: Process[Task, I], write: Sink[Task, W]) {
```

Because in http4s, the `read` parameter is the process who's values get sent to the client and the `write` parameter is the sink to which client requests are sent. It would appear that things are inverted.

Is it just me?

**Comment by bryce-anderson** [comment]
I suppose its not out of the question to support returning multiple types of websocket responses. Just want to read data? How about a WS(mySink). Just want to write data? How about a WS(myProcess)? etc. The problem with that is opening up a can of worms as to what is good enough to get special treatment. It also starts to become more of a DSL than just an API.

**Comment by djspiewak** [comment]
The most idiomatic thing to do here is definitely an `Exchange`.  We do have some experiments going on in the `tcp` module (the new and not-working NIO API) that replace `Exchange` with a `Process` that runs in a reader monad, taking the connection in from the "outside".  This has some advantages and disadvantages, but more important it isn't really part of the public API (the specialized reader monad).  At present, in terms of the public and stable API, `Exchange` is your best bet.

In general, I don't actually look at the names of the values inside of the `Exchange`.  I just pull them out using extractors and remember that the `Process` is the first value while the `Sink` is the second.

`Queue` is generally considered to be a safe and idiomatic structure to use.  `Topic`, not so much, but that's more due to backpressure and scaling concerns than anything else.  `Queue` is very very very useful, and generally doesn't create the design pathologies that you would expect from an inherently mutable construct.  You basically just want to use it at boundaries between concurrently-executing streams.  The chat server is a good example of how it is *supposed* to be used, so I would say that removing it should probably be a non-goal.

I definitely, *definitely* do not recommend throwing `runAsync` into your code anywhere above the level of Blaze!  That's the path to madness.

**Comment by jedesah** [comment]
> The chat server is a good example of how it is supposed to be used, so I would say that removing it should probably be a non-goal.

@djspiewak Yes the chat server requires either a `Topic` or a `Queue`. The problem (which I describe in PR #311) is that a `Queue` should not be required for an echo server or any simple server that transforms an input stream into an output stream.

I just realized that this is the exact same scenario as with [streaming in Remotely](https://github.com/oncue/remotely/pull/79) (don't know why I didn't make this realization earlier). In Remotely we opted for the `Process => Process` pattern.

At this point I am still fairly convinced that the later is the better approach, however it is an important realization that implementing a chat server (something you could very well want to do with Remotely) is a little bit awkward in that model. I am very much open to being convinced that the `Exchange` model is superior.

So at this point I would be interested in exploring how it would be possible to get rid of the `runAsync` in the body of the chat server implementation.

> I definitely, definitely do not recommend throwing runAsync into your code anywhere above the level of Blaze! That's the path to madness.

I generally try very hard (mostly successfully) to avoid calling `run` or anything like that anywhere but at the edge of the world, mostly to avoid the wrath of @runarorama. But in this scenario, I am curious what are the specific disadvantages of doing that and is there a better alternative?

**Comment by djspiewak** [comment]
> At this point I am still fairly convinced that the later is the better approach, however it is an important realization that implementing a chat server (something you could very well want to do with Remotely) is a little bit awkward in that model. I am very much open to being convinced that the Exchange model is superior.

The problem with the `Process => Process` model is that it doesn't give you a "first class" representation of the "writer".  You just produce data, you don't actively write it.  This is an extremely meaningful difference when you're trying to compose exchanges themselves.  For example, we have a function called `Network` which has the following simplified type:

```scala
type PT[A] = Process[Task, A]

// data Address[Node] = Union(Address[Node], Address[Node]) | AtLeast(Int, Set[Node])

forall Node . PT[PT[Exchange[ByteVector, ByteVector]]] => (Node => PT[Exchange[ByteVector, ByteVector]]) => PT[Exchange[ByteVector, (Address[Node], ByteVector)]]
```

So in other words, it's a network management layer which can take incoming connections, request outgoing connections, and merge them all together in a unified interface that we can just "send stuff to".  It's a fantastically useful abstraction, but it is not possible if your fundamental model is `Process => Process`.

> I generally try very hard (mostly successfully) to avoid calling run or anything like that anywhere but at the edge of the world, mostly to avoid the wrath of @runarorama. But in this scenario, I am curious what are the specific disadvantages of doing that and is there a better alternative?

Fairness.  Scheduling.  *Interrupts* (especially interrupts).  Once you do `runAsync`, you cannot ever stop the process.  It's just going to go.  You lose all control, and that basically prevents your cleanup logic from working (in general).

Remember that you can always `drain` and `merge`.  Like so:

```scala
val effects: Process[Task, Unit] = ???    // some process that you want to runAsync
val results: Process[Task, A] = ???     // some random, unrelated process that you need to return

effects.drain merge results       // wye.mergeHaltBoth may also be appropriate
```

This is a super-common idiom in our codebase.  We have some random effects that we need to run, and we don't care about the output, so we just drain + merge with another process and life is good.  You keep scheduling and fairness properties, and interrupts propagate exactly the way you would want.

**Comment by bryce-anderson** [comment]
@djspiewak, thanks for joining the conversation, I've not had my fingers in stream for a while so you're expertise is invaluable. 

I think its now clear that we should avoid a API where one would feel compelled to do a `runAsync` or a `run`. I've though about the 
```scala
whatever.drain ++ whatever2
```
strategy, but that wasn't await of `merge`. (the former wouldn't be of any use for our chat example due to requiring termination of `whatever` before `whatever2` begins its execution) In the end, I think that is rather tricky and not a good fit.

Its sounding more and more like `Exchange` is the right tool for the job.

**Comment by jedesah** [comment]
Okay, so I have thought a lot about the problem these past few days. I think the problem is with `Topic`. `Topic` and `Queue` both very much go against the pull model of scalaz-stream. You push into a `Queue` or `Topic` and this effect unblocks an awaiting stream such as `subscribe` or `dequeue`. `Queue` is invaluable in order to implement a server, because values are coming in no matter what and you need to put them somewhere where a consuming `Process` can eventually retrieve them. However, I think for many other things, the use of `Queue` should be avoided if at all possible, for the exact reason we are encountering here, a `Queue` breaks the chain of steps that each request a value from the previous step (it breaks the pull model).

So in our specific example, even though the output elements of the stream are very much dependent on the input elements, the use of `Topic` breaks the chain and unless the input stream is run, absolutely nothing happens.

So with this said, I believe the solution to the chat server (and to make `Topic` a more useful abstraction) is to add a new function to `Topic` such as `addPublisher(publisher: Process[Task, A])`. The `Topic` will pull from all publishers that have been added to it. I think this can live along side the sink where values are being pushed, so no need to create a new `Topic` like abstraction.

This is what the chat server code would look like with this new method:

    case r @ GET -> Root / "wschat" / name =>

      def frameToMsg(f: WebSocketFrame) = f match {
        case Text(msg, _) => s"$name says: $msg"
        case _            => s"$name sent bad message! Booo..."
      }

      chatTopic.publishOne(s"New user '$name' joined chat")
      .flatMap {_ => WS { in =>

        val receiveMsgs = in.map(frameToMsg).onComplete(Process.eval_(chatTopic.publishOne(s"$name left the chat")))
        chatTopic.addPublisher(receiveMsgs)
        Process.emit(Text(s"Welcome to the chat, $name!")) ++ chatTopic.subscribe.map(Text(_))
      }}

Maybe like me, the sight of `addPublisher` makes you cringe a little bit, but lets not kid ourselves. `Queue` and `Topic` are already very much completely mutable structures and must be manipulated with care.

**Comment by jedesah** [comment]
Out of curiosity, what happens if the `Sink` cannot handle inbound messages fast enough?

**Comment by bryce-anderson** [comment]
Do you mean with the current implementation? Since the `Sink` is driving, there should be no such thing.

**Comment by jedesah** [comment]
@bryce-anderson I am not sure I understand. My understanding is that the `Sink` is a potentially long running operation. In the meantime, the blaze server is receiving messages from the websocket client potentially very quickly. I don't believe the websocket protocol has a notion of backpressure or anything, so what happens if the client overloads the server because the `Sink` is too slow for the frequency of messages being sent by the client?

**Comment by bryce-anderson** [comment]
The `Sink` is what is pulling data from the socket connection. If its not pulling the blaze implementation (which is the only implementation we have right now) is not taking data off the socket. That should cause back pressure on the other end of the wire.

**Comment by jedesah** [comment]
@bryce-anderson Hmm very interesting, that explains things, I guess I need to refresh my memory on socket programming.

**Comment by djspiewak** [comment]
The `Sink` is indeed doing the pulling.  Also, you'll notice that each effect in the `Sink` is sequenced into the stream that results from `to` (or `through`, or `observe`).  In other words, each effect must be completed before more functions (to handle more events) can be synthesized, meaning that the `Sink` automatically slows down if it is having trouble keeping up.  Due to the pull model, this slows down the upstream as well, since the `Sink` itself is in the driver seat.

@jedesah The pull model is insufficient for practical programming, just as the push model is.  The *only* practical stream implementations are a hybrid of both.  You may be interested to know that `wye` and `njoin` *also* break the pull model, since they are effectively an active "puller" in the middle of the stream, batching things up into a queue as they process them.  And why wouldn't they be?  It is not possible to implement concurrency abstractions *without* switching to a push model, for exactly the same reason that it isn't possible to implement a server without the push model *somewhere*.

A `Queue` is the method by which you can "flip" between pull, push and back again.  This is why they are so incredibly ubiquitous in server programming, and why they are far from an anti-pattern.  You certainly shouldn't use them to do anything other than push/pull flipping (though there are some hacks using queues that are currently required due to the stateless nature of `Channel`), but you also shouldn't hesitate to use them whenever your problem conceptually mandates push/pull flipping.  WebSockets almost by definition involve a push/pull flip.

`Topic`, on the other hand, is an abomination.  It really shouldn't be considered part of the public API.  It doesn't support backpressure in any meaningful way, and so any production network of streams that includes a `Topic` is an OOM error waiting to happen (no exaggeration).  `Topic` is very useful internally within scalaz-stream (e.g. `Signal`, which *is* safe to use), but I would not consider it something that anyone should ever use externally.  If you need something that is `Topic`-like, you should use a `TrieMap` (Scala's fast concurrent hash map) and a set of `Queue` instances, glued together with a `mergeN`.

**Comment by jedesah** [comment]
> If its not pulling the blaze implementation (which is the only implementation we have right now) is not taking data off the socket. That should cause back pressure on the other end of the wire.

> it isn't possible to implement a server without the push model somewhere

I may be misunderstanding something here, but I see these two statements as contradictory.


> Topic, on the other hand, is an abomination. It really shouldn't be considered part of the public API.

So I guess the next step should be to rewrite the chat demo in order to avoid the use of `Topic`.

**Comment by djspiewak** [comment]
> I may be misunderstanding something here, but I see these two statements as contradictory.

Blaze itself is certainly propagating backpressure in the manner described.  In fact, what is happening is that backpressure on the source or sink propagates back into Blaze which in turn propagates into TCP itself, which actually has an explicit backpressure mechanism (in the form of backoff and frame size).  There would be more problems if UDP were the underlying protocol, but it's not!  So in other words, it all Just Works™.

Regarding the push model, at *some* level, you need to be able to receive data from the underlying socket.  Or rather, you need to be able to respond to a completed `epoll`.  You don't have the option of saying to the OS "uh, thanks but maybe later" when a poll returns, so what do you do with that data?  You put it into a queue.

Fundamentally, what we're working with are callbacks.  Callbacks, callbacks triggering callbacks, and more callbacks.  Scalaz-stream very explicitly acknowledges this fact with its architecture (which is fundamentally just coroutines within a free monad) and the type signature of `Task.async` (`Task` itself is a just a free monad with an `Async` case, which is literally nothing other than a callback).

The problem of course is that callbacks are *push*, not pull.  Coroutines as a model do a good job of flipping the direction of flow, as it were, but you're still dealing with pushed data at the end of the world.  As you drill down to lower and lower levels of your stack (e.g. Blaze), you need to deal with this explicitly, which is where `async.boundedQueue` comes into play.

> So I guess the next step should be to rewrite the chat demo in order to avoid the use of Topic.

I would recommend that.  It's going to be bulkier, and it's going to slightly obfuscate the elegance of the WebSockets API, but the result will be a better example for people to follow.  In a sense, `Topic`'s only valid use-case is IRC server demos, but the moment you want to "scale" that example into something less trivial, you're going to want something other than `Topic`.

**Comment by jedesah** [comment]
@djspiewak Thanks for the awesome explanation! Things are much clearer to me now on what those two statements mean.

**Comment by jedesah** [comment]
The discussion over on issue #330 made me want to explore another alternative here.

What if instead of asking for an `Exchange`, we provide one:

    case r @ GET -> Root / "wschat" / name =>

        def frameToMsg(f: WebSocketFrame) = f match {
            case Text(msg, _) => s"$name says: $msg"
            case _            => s"$name sent bad message! Booo..."
        }

        chatTopic.publishOne(s"New user '$name' joined chat")
             .flatMap {_ => WS { case Exchange(read, write) =>
                   val receiveMsgs = read.map(frameToMsg).onComplete(Process.eval_(chatTopic.publishOne(s"$name left the chat"))).to(chatTopic.publish).run
                   val responses = Process.emit(Text(s"Welcome to the chat, $name!")) ++ chatTopic.subscribe.map(Text(_))
                   val send = response.to(write).run
                   Applicative[Task].apply2(Task.fork(receiveMsgs), send)
             }}

Unless this is fundamentally incorrect (does not work, particularly worried about the `Task.fork`), I think this is a pretty good compromise. It allows to define an echo server in a pretty straightforward way without requiring a `Queue`. It also lends itself pretty well to defining a chat server IMHO.

**Comment by bryce-anderson** [comment]
All the `.run` in there is a deal breaker in my eyes: its a sign that we're asking the user to drive execution. We want to ask for a computation that _we_ will execute on the users behalf, thus driving effects to the end of the world. Given, the example does return a `Task`, but the difficult part of crunching the `Process`es down to a single delayed computation is really the hard part while the only think left to the backend is to press the 'GO' button.

Regardless of the above, the example you provide is more complex than the current example so the proposed api seems like a step back.

Considering the discussion, I think `Exchange` is going to remain the interface because it allows us to push effects to the boundaries and is more idiomatic with regards to the stream library. @rossabaker, feel free to reopen this if you disagree. Thanks @djspiewak for the enlightening discussion about streams, and thanks @jedesah for the interesting discussion of our current api.

**Comment by jedesah** [comment]
@djspiewak I think you are the one that said that it is unfortunate that `Process.run` is called like so instead of `Process.compile` or something. Do you believe that providing a `Task` instead of a `Process` to the framework diminishes it's ability to orchestrate things?

**Comment by jedesah** [comment]
This is the solution:

    case r @ GET -> Root / "wschat" / name =>

      def frameToMsg(f: WebSocketFrame) = f match {
        case Text(msg, _) => s"$name says: $msg"
        case _            => s"$name sent bad message! Booo..."
      }

      chatTopic.publishOne(s"New user '$name' joined chat")
        .flatMap {_ => WS { in =>

          val receiveMsgs = in.map(frameToMsg).onComplete(Process.eval_(chatTopic.publishOne(s"$name left the chat"))).to(chatTopic.publish)
          val publish = Process.emit(Text(s"Welcome to the chat, $name!")) ++ chatTopic.subscribe.map(Text(_))
          publish merge receiveMsgs.drain
       }}

The `in: Process[Task, WebSocketFrame] => out: Process[Task, WebSocketFrame]` is the easiest model   to wrap your head around and it makes the implementation of an echo server trivial. As for the implementation of a chat server, the above implementation is perfectly reasonable.

**Comment by djspiewak** [comment]
@jedesah To answer your question from earlier, compiling to `Task` and then composing using its concurrency primitives rather than those of `Process` removes scheduling and fairness control from `Process`, which is very very much a bad thing.  It also destroys the ability for `Process` to correctly propagate interrupt semantics.  Basically, your compiled process will just… stop, and without having a chance to run finalizers.  That's not to say that compiling to a monad is a bad thing or that it should *never* be done within the scope of a larger `Process`, but you need to be very very sure that it's the best way to accomplish what you're doing.  Chances are, it's not; and it's *definitely* not a good thing to do when the `Process` you're compiling is long-running.

Regarding your revised signature, I will reiterate that what you're sacrificing with this encoding is the ability to encapsulate the "act of writing" as a separate thing.  If that makes sense for the sorts of problems that it's desirable to use web sockets to address, then by all means, this is the right abstraction to use.  It's not without cost though.  A literal manifestation of this tradeoff is the representation of `out`.  Do you find yourself more often wanting to pipe from one process "into" the output sink?  Or do you want to constructively build up the output?  You can always swap between these modes using a `Queue`, but the default should be right for the majority of use-cases.

You'll notice that you still need a `merge . drain` at the tail end, so it is still very dependent on `async` coordination primitives (like `Queue`).

**Comment by jedesah** [comment]
First of all, thx for your answer on compiling to a Monad and comments on the proposal. I have a few comments inline:

> wanting to pipe from one process "into" the output sink

In the current model, the `Sink` is not used to write to but rather to receive input from the client, which strikes me as odd and I believe awkward to use in the general case (although it works pretty well with `Topic` and the chat server example).

>You can always swap between these modes using a `Queue`, but the default should be right for the majority of use-cases.

Agreed. IMHO the default should be what makes things as simple as possible for a beginner to accomplish simple things like an echo server. I can attest to taking too much time trying to understand the echo server example when I first encountered it even though I am fairly comfortable with FP. If one needs to use a `Queue` or a `Topic` (I know your opinion of `Topic`, but the example still uses it for now) to accomplish more intricate things, then it is reasonable to expect that person to know how to compose the streams in order to make the output stream accomplish all the effects that need to happen.

>You'll notice that you still need a merge . drain at the tail end, so it is still very dependent on async coordination primitives (like Queue).

I am not sure I understand yet what this means exactly, I think it means `merge` uses a `Queue` under the covers. I will have a look at the implementation to try and understand why that is the case but of course will take your word on it for now. But, yes we do need the `async` primitives to accomplish a chat server no matter what because it is inherently stateful, but with the proposed model we do not need to use these primitives to achieve an echo server, or anything that is a pure computation (unlike currently), which in my book is a win.

**Comment by djspiewak** [comment]
> I am not sure I understand yet what this means exactly, I think it means merge uses a Queue under the covers. I will have a look at the implementation to try and understand why that is the case but of course will take your word on it for now. But, yes we do need the async primitives to accomplish a chat server no matter what because it is inherently stateful, but with the proposed model we do not need to use these primitives to achieve an echo server, or anything that is a pure computation (unlike currently), which in my book is a win.

Sorry, my comment was a little indirect.  Whenever you see a `merge . drain` in your code somewhere, it means you have "free effects" that you're just trying to awkwardly fold into `Process`.  This is often very unavoidable, which is part of why I called attention to it earlier in the thread, so I'm in no way saying that it's an anti-pattern.  It's just an interesting flag that indicates the presence of effectful coordination (in this case, `Topic`).

I wasn't actually referencing the use of `Queue` in some of the nondeterminism stuff.  In fact, `merge` is backed by `wye` which does contain a queue, but only in the form of an actor.  `mergeN` contains a `Queue` though.
